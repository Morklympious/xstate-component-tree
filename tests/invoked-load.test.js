"use strict";

const { Machine : createMachine, interpret } = require("xstate");
const trees = require("./util/trees.js");
const component = require("./util/component.js");
const loadAsync = require("./util/async-loader.js");

describe("xstate-component-tree", () => {
    describe("invoked machines using .load", () => {
        it("should support sync .load", async () => {
            const childMachine = createMachine({
                initial : "child",

                states : {
                    child : {
                        meta : {
                            load : () => component("child"),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);
            
            await states(2);
        });
        
        // TODO: broken atm
        // eslint-disable-next-line
        it.skip("should support async .load", async () => {
            const childMachine = createMachine({
                initial : "child",

                states : {
                    child : {
                        meta : {
                            load : loadAsync("child"),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);
            
            await states(3);
        });

        it("should pass child machine context & events to load fns", async () => {
            const childMachine = createMachine({
                initial : "child",

                context : "child context",

                states : {
                    child : {
                        meta : {
                            load : (ctx, event) => ({ ctx, event }),
                        },
                    },
                },
            });

            const testMachine = createMachine({
                initial : "one",

                context : "parent context",

                states : {
                    one : {
                        invoke : {
                            id  : "child",
                            src : childMachine,
                        },

                        meta : {
                            component : component("one"),
                        },
                    },
                },
            });

            const service = interpret(testMachine);

            const states = trees(service);
            
            await states(2);
        });
    });
});
