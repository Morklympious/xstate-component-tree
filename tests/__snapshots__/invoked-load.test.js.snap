// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xstate-component-tree invoked machines using .load should pass child machine context & events to load fns 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": Object {
          "ctx": "child context",
          "event": Object {
            "type": "xstate.init",
          },
        },
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked machines using .load should support async .load 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
]
`;

exports[`xstate-component-tree invoked machines using .load should support sync .load 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "one",
      },
    ],
    "id": "(machine)",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "component": "child",
      },
    ],
    "id": "child",
  },
]
`;
